File: src/sensormodels/HeartRateSensorData.java
package sensormodels;

import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;
import database.LuceneManager;
import org.apache.lucene.document.Document;
import org.apache.lucene.document.Field;
import org.apache.lucene.document.IntPoint;
import org.apache.lucene.document.StringField;
import utils.WebAppConstants;

import java.util.Date;

/**
 * @author Adit Modhvadia
 */
public class HeartRateSensorData implements LuceneStoreModel {

    @SerializedName("sensor_name")
    @Expose
    private String sensorName;
    @SerializedName("timestamp")
    @Expose
    private String timestamp;
    @SerializedName("sensor_data")
    @Expose
    private SensorData sensorData;
    @Expose
    private String formatted_date;

    public String getSensorName() {
        return sensorName;
    }

    public void setSensorName(String sensorName) {
        this.sensorName = sensorName;
    }

    public String getTimestamp() {
        return timestamp;
    }

    public void setTimestamp(String timestamp) {
        this.timestamp = timestamp;
    }

    public SensorData getSensorData() {
        return sensorData;
    }

    public void setSensorData(SensorData sensorData) {
        this.sensorData = sensorData;
    }

    public void setFormattedDate() {
        this.formatted_date = WebAppConstants.inputDateFormat.format(new Date(timestamp));
    }

    public String getFormatted_date() {
        return formatted_date;
    }

    @Override
    public Document getDocument() {
        Document doc = new Document();
//        doc.add(new TextField(LuceneConstants.BPM, String.valueOf(sensorData.getSensorData().getBpm()), Field.Store.YES));
        doc.add(new IntPoint(LuceneManager.LuceneConstants.BPM, this.getSensorData().getBpm()));
        doc.add(new StringField(LuceneManager.LuceneConstants.SENSOR_NAME, this.getSensorName(), Field.Store.YES));
        doc.add(new StringField(LuceneManager.LuceneConstants.FORMATTED_DATE, this.getFormatted_date(), Field.Store.YES));
//         use a string field for timestamp because we don't want it tokenized
        doc.add(new StringField(LuceneManager.LuceneConstants.TIMESTAMP, this.getTimestamp(), Field.Store.YES));
        return doc;
    }

    public static class SensorData {

        public SensorData() {
        }

        @SerializedName("bpm")
        @Expose
        private Integer bpm;

        public Integer getBpm() {
            return bpm;
        }

        public void setBpm(Integer bpm) {
            this.bpm = bpm;
        }

    }
}

File: src/main/java/sensormodels/HeartRateSensorData.kt
package sensormodels

import com.google.gson.annotations.Expose
import com.google.gson.annotations.SerializedName
import database.LuceneManager
import org.apache.lucene.document.Document
import org.apache.lucene.document.Field
import org.apache.lucene.document.IntPoint
import org.apache.lucene.document.StringField
import org.bson.codecs.pojo.annotations.BsonIgnore
import sensormodels.HeartRateSensorData
import sensormodels.store.models.FileStoreModel
import sensormodels.store.models.LuceneStoreModel
import sensormodels.store.models.MongoStoreModel
import sensormodels.store.models.MySqlStoreModel
import utils.WebAppConstants
import java.io.File
import java.sql.PreparedStatement
import java.sql.SQLException
import java.util.*

class HeartRateSensorData : MongoStoreModel, LuceneStoreModel, FileStoreModel, MySqlStoreModel {
    @SerializedName("sensor_name")
    @Expose
    var sensorName: String? = null

    @SerializedName("timestamp")
    @Expose
    var timestamp: String? = null

    @SerializedName("sensor_data")
    @Expose
    var sensorData: SensorData? = null

    @SerializedName("formatted_date")
    @Expose
    var formattedDate: String? = null
        private set

    @BsonIgnore
    private var file: File? = null
    override fun setFormattedDate() {
        formattedDate = WebAppConstants.inputDateFormat.format(Date(timestamp))
    }

    override fun getStartTime(): String {
        return timestamp!!
    }

    @BsonIgnore
    override fun getDocument(): Document {
        val doc = Document()
        //        doc.add(new TextField(LuceneConstants.BPM,
        // String.valueOf(sensorData.getSensorData().getBpm()), Field.Store.YES));
        doc.add(IntPoint(LuceneManager.LuceneConstants.BPM, sensorData!!.bpm!!))
        doc.add(
            StringField(
                LuceneManager.LuceneConstants.SENSOR_NAME, sensorName, Field.Store.YES
            )
        )
        doc.add(
            StringField(
                LuceneManager.LuceneConstants.FORMATTED_DATE,
                formattedDate,
                Field.Store.YES
            )
        )
        //         use a string field for timestamp because we don't want it tokenized
        doc.add(
            StringField(
                LuceneManager.LuceneConstants.TIMESTAMP, timestamp, Field.Store.YES
            )
        )
        return doc
    }

    @BsonIgnore
    override fun getMongoCollectionName(): String {
        return MONGO_COLLECTION_NAME
    }

    @BsonIgnore
    override fun getClassObject(): Class<HeartRateSensorData> {
        return HeartRateSensorData::class.java
    }

    @BsonIgnore
    override fun getTableName(): String {
        return MY_SQL_TABLE_NAME
    }

    @BsonIgnore
    override fun getCreateTableQuery(): String {
        return ("CREATE TABLE ${this.tableName}(timestamp VARCHAR(30) ,  formatted_date VARCHAR(10) ,  sensor_name CHAR (25),  bpm INTEGER)")
    }

    @BsonIgnore
    override fun getInsertIntoTableQuery(): String {
        return (" insert into ${this.tableName} (timestamp, formatted_date, sensor_name,bpm) values (?, ?, ?, ?)")
    }

    @BsonIgnore
    @Throws(SQLException::class)
    override fun fillQueryData(preparedStmt: PreparedStatement) {
        preparedStmt.apply {
            setString(1, timestamp)
            setString(2, formattedDate)
            setString(3, sensorName)
            setInt(4, sensorData!!.bpm!!)
        }
    }

    @BsonIgnore
    override fun getFileName(): String {
        return FILE_NAME
    }

    @BsonIgnore
    override fun getFile(): File {
        return file!!
    }

    override fun setFile(file: File) {
        this.file = file
    }

    class SensorData {
        @SerializedName("bpm")
        @Expose
        var bpm: Int? = null
    }

    companion object {
        @BsonIgnore
        val MY_SQL_TABLE_NAME = "HeartRateSensorData"

        @BsonIgnore
        val MONGO_COLLECTION_NAME = "HeartRateSensorData"

        @BsonIgnore
        val FILE_NAME = "HeartRate"
    }
}